internal/mage/util/engine.go

```diff
@@ -230,8 +239,68 @@ func devEngineContainer(c *dagger.Client, arch string, version string, opts ...D
 		WithNewFile(engineEntrypointPath, dagger.ContainerWithNewFileOpts{
 			Contents:    engineEntrypoint,
 			Permissions: 0o755,
+		})
+	return container.WithEntrypoint([]string{"dagger-entrypoint.sh"})
+}
+
+func devEngineContainerWithGPUSupport(c *dagger.Client, arch string, version string, opts ...DevEngineOpts) *dagger.Container {
+	engineConfig, err := getConfig(opts...)
+	if err != nil {
+		panic(err)
+	}
+	engineEntrypoint, err := getEntrypoint(opts...)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("arch", arch)
+
+	container := c.Container(dagger.ContainerOpts{Platform: dagger.Platform("linux/" + arch)}).
+		From("ubuntu:"+ubuntuVersion).
+		WithEnvVariable("DEBIAN_FRONTEND", "noninteractive").
+		WithExec([]string{"apt-get", "update"}).
+		WithExec([]string{
+			"apt-get", "install", "-y",
+			"iptables", "git", "dnsmasq-base", "network-manager",
+			"gpg", "curl",
 		}).
-		WithEntrypoint([]string{"dagger-entrypoint.sh"})
+		WithFile("/usr/local/bin/runc", runcBin(c, arch), dagger.ContainerWithFileOpts{
+			Permissions: 0o700,
+		}).
+		WithFile("/usr/local/bin/buildctl", buildctlBin(c, arch)).
+		WithFile("/usr/local/bin/"+shimBinName, shimBin(c, arch)).
+		WithFile("/usr/local/bin/"+engineBinName, engineBin(c, arch, version)).
+		WithFile("/usr/local/bin/"+daggerBinName, daggerBin(c, arch, version)).
+		WithFile(consts.GoSDKEngineContainerTarballPath, goSDKImageTarBall(c, arch)).
+		WithDirectory(filepath.Dir(consts.PythonSDKEngineContainerModulePath), pythonSDK(c)).
+		WithDirectory("/usr/local/bin", qemuBins(c, arch)).
+		WithDirectory("/", cniPlugins(c, arch, true)).
+		WithDirectory(EngineDefaultStateDir, c.Directory()).
+		WithNewFile(engineTomlPath, dagger.ContainerWithNewFileOpts{
+			Contents:    engineConfig,
+			Permissions: 0o600,
+		}).
+		WithNewFile(engineEntrypointPath, dagger.ContainerWithNewFileOpts{
+			Contents:    engineEntrypoint,
+			Permissions: 0o755,
+		}).
+		With(nvidiaSetup)
+
+	return container.WithEntrypoint([]string{"dagger-entrypoint.sh"})
+}
+
+// install nvidia-container-toolkit in the container
+func nvidiaSetup(ctr *dagger.Container) *dagger.Container {
+	return ctr.
+		With(shellExec(`curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg`)).
+		With(shellExec(`curl -s -L https://nvidia.github.io/libnvidia-container/experimental/"$(. /etc/os-release;echo $ID$VERSION_ID)"/libnvidia-container.list | sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | tee /etc/apt/sources.list.d/nvidia-container-toolkit.list`)).
+		With(shellExec(`apt-get update && apt-get install -y nvidia-container-toolkit`))
+}
+
+func shellExec(cmd string) dagger.WithContainerFunc {
+	return func(ctr *dagger.Container) *dagger.Container {
+		return ctr.WithExec([]string{"sh", "-c", cmd})
+	}
```

Comment:
I would like to see more code here & fewer scripts. FWIW:
- https://pkg.go.dev/dagger.io/dagger#Client.HTTP
- https://pkg.go.dev/dagger.io/dagger#Container.WithFile
- https://pkg.go.dev/dagger.io/dagger#Container.WithNewFile

It's OK to leave this for a follow-up.